servant/src/Servant/Test/ComprehensiveAPI.hs

```diff
@@ -15,38 +15,64 @@ import           Servant.Types.SourceT
-  "params"           :> QueryParams "foo" Int :> GET :<|>
-  "flag"             :> QueryFlag "foo" :> GET :<|>
-  "remote-host"      :> RemoteHost :> GET :<|>
-  "req-body"         :> ReqBody '[JSON] Int :> GET :<|>
-  "req-body-lenient" :> ReqBody' '[Lenient] '[JSON] Int :> GET :<|>
-  "res-headers"      :> Get '[JSON] (Headers '[Header "foo" Int] NoContent) :<|>
-  "foo"              :> GET :<|>
-  "vault"            :> Vault :> GET :<|>
-  "post-no-content"  :> Verb 'POST 204 '[JSON] NoContent :<|>
-  "post-int"         :> Verb 'POST 204 '[JSON] Int :<|>
-  "streaming"        :> StreamBody NetstringFraming JSON (SourceT IO Int) :> Stream 'GET 200 NetstringFraming JSON (SourceT IO Int) :<|>
-  "named-context"    :> WithNamedContext "foo" '[] GET :<|>
-  "capture-all"      :>  CaptureAll "foo" Int :> GET :<|>
-  "summary"          :> Summary "foo" :> GET :<|>
-  "description"      :> Description "foo" :> GET :<|>
-  "empty-api"        :> EmptyAPI
+    ComprehensiveAPIWithoutStreamingOrRaw'
+    (EmptyEndpoint :<|> StreamingEndpoint)
 
 comprehensiveAPIWithoutRaw :: Proxy ComprehensiveAPIWithoutRaw
 comprehensiveAPIWithoutRaw = Proxy
+
+type ComprehensiveAPIWithoutStreaming =
+    ComprehensiveAPIWithoutStreamingOrRaw'
+    (EmptyEndpoint :<|> RawEndpoint)
+
+comprehensiveAPIWithoutStreaming :: Proxy ComprehensiveAPIWithoutStreaming
+comprehensiveAPIWithoutStreaming = Proxy
+
+-- | @:: API -> API@, so we have linear structure of the API.
+type ComprehensiveAPIWithoutStreamingOrRaw' endpoint =
+    GET
+    :<|> "get-int"          :> Get '[JSON] Int
+    :<|> "capture"          :> Capture' '[Description "example description"] "foo" Int :> GET
+    :<|> "header"           :> Header "foo" Int :> GET
+    :<|> "header-lenient"   :> Header' '[Required, Lenient] "bar" Int :> GET
+    :<|> "http-version"     :> HttpVersion :> GET
+    :<|> "is-secure"        :> IsSecure :> GET
+    :<|> "param"            :> QueryParam "foo" Int :> GET
+    :<|> "param-lenient"    :> QueryParam' '[Required, Lenient] "bar" Int :> GET
+    :<|> "params"           :> QueryParams "foo" Int :> GET
+    :<|> "flag"             :> QueryFlag "foo" :> GET
+    :<|> "remote-host"      :> RemoteHost :> GET
+    :<|> "req-body"         :> ReqBody '[JSON] Int :> GET
+    :<|> "req-body-lenient" :> ReqBody' '[Lenient] '[JSON] Int :> GET
+    :<|> "res-headers"      :> Get '[JSON] (Headers '[Header "foo" Int] NoContent)
+    :<|> "foo"              :> GET
+    :<|> "vault"            :> Vault :> GET
+    :<|> "post-no-content"  :> Verb 'POST 204 '[JSON] NoContent
+    :<|> "post-int"         :> Verb 'POST 204 '[JSON] Int
+    :<|> "named-context"    :> WithNamedContext "foo" '[] GET
+    :<|> "capture-all"      :> CaptureAll "foo" Int :> GET
+    :<|> "summary"          :> Summary "foo" :> GET
+    :<|> "description"      :> Description "foo" :> GET
+    :<|> "alternative"      :> ("left" :> GET :<|> "right" :> GET)
```

Comment:
this case is added to make a (small) tree, as otherwise the structure of `ComprehensiveAPI` is linear.